### PR TITLE
feat: implement tokens metadata storing (improving Jamie's 222 pr) 

### PR DESCRIPTION
**Pull request by dmytro-zaharnytskyi**
# PR Description: API Key Management with SQLite Metadata Store

## Summary
This PR introduces a robust system for managing named API keys. Instead of relying on Kubernetes Secrets for metadata (which risks "Secret Explosion" or complexity with aggregation), we have implemented a **SQLite-based metadata store**. This separates the concern of *token generation* (handled via Kubernetes Service Accounts) from *token metadata persistence* (names, creation dates, owner).

### 1. SQLite Metadata Storage
*   **Architecture:** Replaced the proposed Kubernetes Secret storage with a local SQLite database.
*   **Why:** Provides efficient querying, sorting, and persistence for named tokens without burdening the Kubernetes API server or dealing with etcd size limits.
*   **Schema:** Stores `id` (JTI), `username`, `name`, `namespace`, `creation_date`, and `expiration_date`.

### 2. Package Restructuring & Separation of Concerns
*   **New `internal/api_keys` Package:** Dedicated package for managing the lifecycle of named API keys, including the SQLite store, service layer, and HTTP handlers.
*   **Refactored `internal/token` Package:** Streamlined to focus solely on Kubernetes Service Account interactions (generating and reviewing tokens). It no longer concerns itself with metadata persistence.
*   **Interface Driven:** The `api_keys.Service` relies on a `TokenManager` interface, allowing for easier testing and future backend swaps.

### 3. Namespace Isolation
*   **Security:** Implemented strict namespace filtering. Token retrieval and revocation are scoped to the user's tier-bound namespace.
*   **Implementation:** The `Store` now enforces `WHERE namespace = ?` clauses on all reads and deletes to prevent cross-tenant data leakage.

### 4. API Endpoints
Implemented a RESTful interface for API key management under `/v1/api-keys`:
*   `POST /v1/api-keys`: Create a new named token.
*   `GET /v1/api-keys`: List all active and expired keys for the authenticated user.
*   `GET /v1/api-keys/:id`: Retrieve details for a specific key.
*   `DELETE /v1/api-keys/:id`: Revoke a specific key (removes metadata; underlying SA token remains valid until expiration or full revocation).
*   `DELETE /v1/tokens`: Revoke **all** tokens (recreates the underlying Service Account).

### 5. Quality & Reliability
*   **Linter Compliance:** Addressed all `golangci-lint` issues including `contextcheck`, `godot` (comments), `goconst`, and `errcheck`.
*   **Testing:** Added comprehensive unit tests covering:
    *   Store CRUD operations.
    *   Cross-namespace isolation (ensuring User A in Namespace X cannot see User A's tokens in Namespace Y).
    *   Handler edge cases and type safety.

## Metadata Retrieval Flow

**New Logic:**
1.  **Request:** User requests their API keys (`GET /v1/api-keys`).
2.  **Context:** Middleware extracts the user's identity and tier.
3.  **Namespace Resolution:** The service determines the user's target namespace based on their tier.
4.  **Query:** The SQLite store queries `tokens` filtering by `username` AND `namespace`.
5.  **Response:** Returns a JSON list of named tokens with their status (`active` or `expired`).

## Summary by CodeRabbit

* **New Features**
  * Added API Key management: create, list, and retrieve named API keys via /v1/api-keys; metadata persisted to on-disk DB.
  * Tokens now expose JTI and issuance metadata.

* **Behavior Changes**
  * DELETE /v1/tokens now revokes ephemeral tokens for the authenticated user; API keys are managed separately.

* **Infrastructure**
  * Persistent storage added for API keys (PVC and DB path set).

* **Documentation**
  * Expanded docs and examples for ephemeral tokens and API key workflows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
### General feedback

Solid work, thanks for pushing this forward.

For this PR, I would really like to **focus the scope on “managed tokens/keys”** and avoid mixing in deny-list / model-access concerns. That should make it easier to ship this incrementally.

Right now the `deny-list` logic is wired around securing **maas-api endpoints**, but what we actually need is a deny-list that applies to **model access**. It is a follow-up piece of work rather than something we should solve in this PR.

From a cursory review, maybe we can split the behaviour into explicit endpoints instead of one endpoint with conditional semantics:

* `POST /v1/issue` (old `/v1/tokens`): Returns a JWT only and provides auto-refresh. No persistence.
* `POST /v1/api-keys` (metadata required): Returns JWT **and** persists metadata (managed key).
* `GET /v1/api-keys`: Lists metadata for managed keys owned by the current user (maybe partial objects: `id` + `name`, paginated for UX).
* `GET /v1/api-keys/{id}`: Returns full metadata for a specific managed key.

One more important design point: we should use the JWT `jti` as the primary identifier of a key, not a DB-specific surrogate `id`. That means:

* `jti` is stored in the DB.
* `jti` is exposed as a top-level field in request/response payloads (convenient too)
* Any future deny-list / revocation logic can key off `jti` consistently.

> [!IMPORTANT]
> For now, I would  **defer `DELETE` for a single key**. We are using the same issuer for JWTs, so the revoke story needs to handle both ephemeral and stored tokens via a shared deny-list. Let's design that holistically (models + keys).

---

### Technical improvements

* Treat **revocation** and **expiration** as two distinct states:  
  * Expiration = derived directly from the JWT (`exp`), the token naturally becomes invalid.  
  * Revocation = explicit deny-list entry blocking access even if `exp` hasn’t passed (but this is out of scope for this PR)
* If we want to persist timestamps, read them from the issued JWT instead of maintaining a second source of truth.
* DB operations tied to the token lifecycle should be **transactional**

---

### Follow-up work

* **Pluggable database**  
  * SQLite is fine for now, but it still comes with operational overhead. An in-memory backend could help for dev/tests or small deployments.
* **Deny-list for model access**  
  * Design deny-list semantics around model access

And one more thing: let’s avoid adding more bash-based tests. They quickly become maintenance overhead; logic like this should preferably live in Go or our e2e pytests.

File: maas-api/internal/token/manager.go [hunk 1]
```diff
@@ -97,13 +114,23 @@ func (m *Manager) RevokeTokens(ctx context.Context, user *UserContext) error {
 	_, err = m.serviceAccountLister.ServiceAccounts(namespace).Get(saName)
 	if errors.IsNotFound(err) {
 		log.Printf("Service account %s not found in namespace %s, nothing to revoke", saName, namespace)
+		// Still try to mark secrets as expired even if SA doesn't exist
```
Comment (jland-redhat) [inline maas-api/internal/token/manager.go]: Same comment as above

File: maas-api/internal/token/manager.go [hunk 2]
```diff
@@ -117,6 +144,60 @@ func (m *Manager) RevokeTokens(ctx context.Context, user *UserContext) error {
...
+	// If it is a User token (not SA), we should allow it (Bootstrap/Admin access) and skip DB check.
+	if !strings.HasPrefix(userCtx.Username, "system:serviceaccount:") {
+		log.Printf("Allowing non-SA token for user: %s", userCtx.Username)
+		return userCtx, nil
+	}
+
+	// 3. Check deny list (DB)
+	// Compute hash of the token
+	tokenHash := HashToken(token)
+	log.Printf("Checking token hash in DB for user: %s, hash: %s", userCtx.Username, tokenHash[:16]+"...")
```
Comment (jland-redhat) [inline maas-api/internal/token/manager.go]: So I think we are just checking the has value here if I am understanding correctly.

I think we wanted to throw a deny if the tokens had the same name so probably need to modify this check to look for that.

File: maas-api/cmd/main.go [hunk 3]
```diff
@@ -124,16 +128,65 @@ func configureSATokenProvider(ctx context.Context, cfg *config.Config, router *g
...
+	cleanup := func() {
+		if err := store.Close(); err != nil {
+			log.Printf("Failed to close token store: %v", err)
+		}
+	}
+	return cleanup, store
+}
+
+// startTokenExpirationCleanup runs a background job that periodically marks expired tokens
+// It runs every 5 minutes and stops when the context is cancelled
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply maas-api/cmd/main.go]: Expiry can be derived from token itself, or we can use those timestamps fields in the query. Having extra status column for that and background job sounds like an overkill to me.

**Comment by bartoszmajsak** [comment]
> One question this actually brought up for me around "SQLite is fine for now, but it still comes with operational overhead. An in-memory backend could help for dev/tests or small deployments":
> Is the sql instance not built into our image? I am actually having a hard time seeing where our DB is being deployed, is go somehow actually starting an instance of the db?


SQLite is embedded in the Go binary, there’s no separate DB service. The "deployment" part is the PVC: the app just opens a SQLite file on that volume, which is why we currently need a StatefulSet + PVC to keep storage stable across restarts.

That's the operational overhead I was referring to: even though it is "just SQLite", we still manage stateful pods, storage classes, backups, etc.

I agree that long-term we probably don't want to stick with SQLite and will need something more scalable. To make that easier (and give cluster admins more choice, similar to Limitador), we should treat storage as pluggable choice, so I was thinking:

* SQLite on PVC for "production-ready" deployments (emphasis on "")
* an in-memory backend (or SQLite :memory:) for dev/tests or tiny installs

On the k8s side we can wire this up via kustomize overlays as a first step (different overlays for in-memory vs PVC-backed, and later Postgres/whatever).

**Comment by bartoszmajsak** [comment]
> Another comment is that I do think we are going to want a "delete" endpoint that actually deletes the metadata which would be separate from expiring a token

@Jaland Not sure, let's defer it. We do not treat "metadata" as a subresource right now.

File: smoke-test/README.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+#### 2. Token Functionality
+- Token can be used to access protected endpoints
+- GET request to `/v1/models` succeeds with status 200
+
+#### 3. Invalid Token Handling
+- Invalid/malformed tokens return 401 Unauthorized
+- API properly rejects unauthenticated requests
+
+#### 4. Kubernetes Secret Metadata
```
[the remark this answers is not in the document]
Comment (dmytro-zaharnytskyi) [reply smoke-test/README.md]: yep, leftover, will check

File: maas-api/cmd/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply maas-api/cmd/main.go]: YAGNI if you ask me :)

File: maas-api/internal/token/manager.go [hunk 5]
```diff
@@ -97,13 +125,23 @@ func (m *Manager) RevokeTokens(ctx context.Context, user *UserContext) error {
 	_, err = m.serviceAccountLister.ServiceAccounts(namespace).Get(saName)
 	if errors.IsNotFound(err) {
 		log.Printf("Service account %s not found in namespace %s, nothing to revoke", saName, namespace)
+		// Still try to delete token metadata even if SA doesn't exist
+		if err := m.store.DeleteTokensForUser(ctx, namespace, user.Username); err != nil {
+			log.Printf("Failed to delete tokens for user %s: %v", user.Username, err)
+		}
```
Comment (bartoszmajsak) [inline maas-api/internal/token/manager.go]: If we have new `api_keys` service/manager we could do this call there instead of mixing two responsibilities here.

File: maas-api/internal/token/types.go [hunk 6]
```diff
@@ -26,44 +25,31 @@ type Duration struct {
 	time.Duration
 }
 
-func (d *Duration) MarshalJSON() ([]byte, error) {
+func (d Duration) MarshalJSON() ([]byte, error) {
```
Comment (bartoszmajsak) [inline maas-api/internal/token/types.go]: Would it make sense to do `nil` check instead and keep pointer receiver?

File: maas-api/cmd/main.go [hunk 7]
```diff
@@ -124,17 +123,42 @@ func configureSATokenProvider(ctx context.Context, cfg *config.Config, router *g
...
 		serviceAccountInformer.Lister(),
+		store,
 	)
 	tokenHandler := token.NewHandler(cfg.Name, manager)
+	// Create reviewer with audience to properly validate Service Account tokens
+	reviewer := token.NewReviewerWithAudience(clusterConfig.ClientSet, cfg.Name+"-sa")
+
+	// Protect model listing endpoints with token validation
+	router.GET("/models", tokenHandler.ExtractUserInfo(reviewer), modelsHandler.ListModels)
+	v1Routes.GET("/models", tokenHandler.ExtractUserInfo(reviewer), modelsHandler.ListLLMs)
```
Comment (bartoszmajsak) [inline maas-api/cmd/main.go]:
```suggestion
	v1Routes.GET("/v1/models", tokenHandler.ExtractUserInfo(reviewer), modelsHandler.ListLLMs)
```

File: maas-api/internal/token/reviewer.go [hunk 8]
```diff
@@ -22,18 +23,60 @@ func NewReviewer(clientset kubernetes.Interface) *Reviewer {
 	}
 }
 
+// NewReviewerWithAudience creates a new Reviewer instance with audience
+func NewReviewerWithAudience(clientset kubernetes.Interface, audience string) *Reviewer {
+	return &Reviewer{
+		clientset: clientset,
+		audience:  audience,
+	}
+}
```
[the remark this answers is not in the document]
Comment (dmytro-zaharnytskyi) [reply maas-api/internal/token/reviewer.go]: I see it as tokens created with an audience should be validated with that audience. Yep, we can simplify approach by merging this into one function with an optional audience parameter.

Ahh.. agree with you, will simplify

**Review by jland-redhat — CHANGES_REQUESTED** [review]

File: maas-api/internal/token/types.go [hunk 9]
```diff
@@ -26,44 +25,31 @@ type Duration struct {
...
-var allowedUnits = regexp.MustCompile(`^(\d+(?:\.\d+)?[hms])+$`)
-
 func (d *Duration) UnmarshalJSON(b []byte) error {
 	var v any
 	if err := json.Unmarshal(b, &v); err != nil {
 		return err
 	}
-
 	switch value := v.(type) {
+	case float64:
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply maas-api/internal/token/types.go]: But we already [have it](https://github.com/opendatahub-io/maas-billing/blob/3e625bb3900b2a1b2c88899e48202efa6c091b06/maas-api/internal/token/types.go#L54-L59). I think what changed here is the move of validation logic, which was done as part of unmarshalling, now we have it in two places duplicated and we are loosing consistency.

File: maas-api/internal/api_keys/service.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+// It recreates the Service Account (invalidating all tokens) and marks API key metadata as expired.
+func (s *Service) RevokeAll(ctx context.Context, user *token.UserContext) error {
+	// Revoke in K8s (recreate SA) - this invalidates all tokens
+	namespace, err := s.tokenManager.RevokeTokens(ctx, user)
+	if err != nil {
+		return fmt.Errorf("failed to revoke tokens in k8s: %w", err)
+	}
+
+	// Mark API key metadata as expired (preserves history)
+	if err := s.store.MarkTokensAsExpiredForUser(ctx, namespace, user.Username); err != nil {
```
Comment (bartoszmajsak) [inline maas-api/internal/api_keys/service.go]: Shouldn't we delete revoked tokens? Why do we want to keep them for the user?

File: maas-api/internal/api_keys/types.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package api_keys
+
+// NamedToken represents metadata for a single token.
+type NamedToken struct {
```
Comment (bartoszmajsak) [inline maas-api/internal/api_keys/types.go]: Is it token? maybe `ApiKey`?

**Review by bartoszmajsak — COMMENTED** [review]
Mostly minor comments, but we should also plan follow-ups:

- ensure consistency of new operations
- pluggable storage
- pagination
- passive/active cleanup of expired keys

Anything else @dmytro-zaharnytskyi @jland-redhat ?

File: maas-api/internal/api_keys/store.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+	rows, _ := result.RowsAffected()
+	log.Printf("Marked %d tokens as expired for user %s", rows, username)
+	return nil
+}
+
+// DeleteToken is deprecated and non-functional - kept for interface compatibility.
+// Single token deletion is not supported in the initial release.
+// Use MarkTokensAsExpiredForUser to revoke all tokens for a user.
+func (s *Store) DeleteToken(ctx context.Context, namespace, username, jti string) error {
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply maas-api/internal/api_keys/store.go]: And if we don't need it - GetNsForUser can go away too.

File: maas-api/internal/api_keys/handler.go [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+		}
+		log.Printf("Failed to get api key: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve API key"})
+		return
+	}
+
+	c.JSON(http.StatusOK, tok)
+}
+
+func (h *Handler) RevokeAPIKey(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (dmytro-zaharnytskyi) [reply maas-api/internal/api_keys/handler.go]: Yep, I left it intentionally, to come back to it after release. But you are right, better to keep the code clean and then add it from scratch. Removing RevokeAPIKey

File: .gitignore [hunk 14]
```diff
@@ -19,6 +19,12 @@ coverage.out
 coverage.html
 bin/
 
+# Go binaries
+maas-api/main
+maas-api/*/main
+*.exe
+*.exe~
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply .gitignore]: Maybe we can avoid adding AI ideas that are unrelated to the scope of this work?

File: maas-api/internal/token/handler.go [hunk 15]
```diff
@@ -9,20 +12,25 @@ import (
 	"github.com/gin-gonic/gin"
 )
 
+type TokenManager interface {
+	GenerateToken(ctx context.Context, user *UserContext, expiration time.Duration, name string) (*Token, error)
+	ValidateToken(ctx context.Context, token string, reviewer *Reviewer) (*UserContext, error)
+}
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply maas-api/internal/token/handler.go]: let's be pragmatic and avoid more technical debt.

- we don't need strict interfaces in Go unless really necessary, moreover we do not ensure that implementations are adhering to them anywhere in the code, so we are in fact not enforcing the contract
- the validation can come as separate PR - removing code is quick. I am also not sure what we are really validating here more than we can do in Auth policy we have in place

**Review by bartoszmajsak — COMMENTED** [review]
Functionally it is probably fine, but we have accidental complexity, which makes it harder to review and will probably increase maintenance cost moving forward. I am not sure if we feel good about merging it and do immediate follow-ups or address it in this PR before merging? Relevant parts:

* https://github.com/opendatahub-io/maas-billing/pull/232#discussion_r2589396721
* https://github.com/opendatahub-io/maas-billing/pull/232#discussion_r2589370496
* https://github.com/opendatahub-io/maas-billing/pull/232#discussion_r2589362854
* https://github.com/opendatahub-io/maas-billing/pull/232#discussion_r2589428965
* https://github.com/opendatahub-io/maas-billing/pull/232#discussion_r2589364434

File: maas-api/internal/api_keys/store.go [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	// 2. Migrate existing tables: add description column if it doesn't exist
+	// SQLite doesn't support "ALTER TABLE ... ADD COLUMN IF NOT EXISTS", so we check first
+	var columnExists int
+	checkColumnQuery := `
+	SELECT COUNT(*) FROM pragma_table_info('tokens') WHERE name='description';
+	`
+	err := s.db.QueryRowContext(ctx, checkColumnQuery).Scan(&columnExists)
+	if err == nil && columnExists == 0 {
```
[the remark this answers is not in the document]
Comment (dmytro-zaharnytskyi) [reply maas-api/internal/api_keys/store.go]: Oh, I think it will be put in a far far backlog but why not. @jland-redhat if you also agree on that - can you please create a jyra for that?

**Comment by jland-redhat** [comment]
lgtm
